### ISSUE TITLE
Type arguments used in return types of argument functions are not correctly inferred when wrapped in objects

### ISSUE THREAD

**Issue by chelproc** [microsoft/TypeScript#42379]
# Bug Report

### 🔎 Search Terms

* Return types
* Type inference

### 🕗 Version & Regression Information

* This is a bug
* This happens on all of the latest versions usable in the playground (4.1.3, 4.2.0-beta, even nightly)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.1.3&ssl=1&ssc=1&pln=34&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzBwFsAjLVEYAHgBUA+AChxICsAueAbwCh4kBGDg0RYAHpQAKsKEQ4BnDDHIBzAJTwAvHXg0A3LyQAmIeUQgYcYFJgyONdVvhoA1qhwB3VPoC+qji7dPfW4AehD4AGEoOQR+bkJSckoGHj5EQXhhMUlpIgdtfgAaA0RjTNNzS2sZfK5vQvgwp1RXD1RuX2CmqJj4Q3jiMgpgFJKMrPErXI4oVABPWqKSsoYKixybPM1tTnrG8NRkUnMO1S7wnoQAZgHE4dG08cXitJW1qtza3Yamw+OYU7BUCQWAIFDobB4eAxAAO0gwlFojAMzBYGU4AiEIkm1VkzVanlqNHg9RRrDKGNKJlQZnWU02dlqATaJOKfnxgS83FCF2iCAALNxYfDkgZKeNsRsatt4PxWWKjNTaR9Nl9vKz9vA-iQTmduEA)

### 💻 Code

```ts
declare function combined<T>(obj: {
  f1: (fixedParam: string) => T;
  f2: (inferredParam: T) => unknown;
}): unknown;

// Case 1
combined({
  f1: (fixedParam) => 1,
  f2: (inferredParam) => {}, // unknown
});

// Case 2
combined({
  f1: (fixedParam: any) => 1,
  f2: (inferredParam) => {}, // number
});

// Case 3
combined({
  f1: () => 1,
  f2: (inferredParam) => {}, // number
});

declare function separated<T>(
  obj1: { f1: (fixedParam: unknown) => T },
  obj2: { f2: (inferredParam: T) => unknown },
): unknown;

// Case 4
separated(
  { f1: (fixedParam) => 1 },
  { f2: (inferredParam) => {} }, // number
);
```

### 🙁 Actual behavior

There is a function with a type argument `T`. It requires an object as an argument that has two properties:
* A function that returns `T`
* A function that requires `T` as an argument.

If the first function requires any argument that must be inferred from the signature, TypeScript will fail to infer `T` (Case 1). 

This behavior does not occur when:
* `fixedParam` is explicitly typed (Case 2) or not used (Case 3).
or
* Two inner functions are in separate arguments of outer functions (Case 4).

### 🙂 Expected behavior

All `inferredParam` should be inferred as `number`. This feature is useful when struggling with handling complex data structure in React.

**Comment by RyanCavanaugh** [comment]
The inference algorithm doesn't support this sort of setup (and can't be "fixed" to do so). See #30134

**Comment by jcalz** [comment]
Fixed by #48538, right?
